File: задача4_решение.md/Externallibs/DatabaseSavedecorator.cs
public class DatabaseSaverDecorator : IDatabaseSaver
{
    private readonly IDatabaseSaver _wrappedSaver;
    private readonly MailSender _mailSender;
    private readonly CacheUpdater _cacheUpdater;

    public DatabaseSaverDecorator(IDatabaseSaver saver, MailSender mailSender, CacheUpdater cacheUpdater)
    {
        _wrappedSaver = saver;
        _mailSender = mailSender;
        _cacheUpdater = cacheUpdater;
    }

    public void SaveData(object data)
    {
        _wrappedSaver.SaveData(data);
        _mailSender.Send("admin@example.com");
        _cacheUpdater.UpdateCache();
    }
}

// Клиентский код
public class DatabaseSaverClient
{
    public void Main(bool b)
    {
        var saver = new DatabaseSaverDecorator(
            new DatabaseSaver(),
            new MailSender(),
            new CacheUpdater()
        );
        DoSmth(saver);
    }

    private void DoSmth(IDatabaseSaver saver)
    {
        saver.SaveData(null);
    }
}

File: задача4_решение.md/DatabaseSavedecorator.py
class DatabaseSaverDecorator:
    def __init__(self, saver, mail_sender, cache_updater):
        self._saver = saver
        self._mail_sender = mail_sender
        self._cache_updater = cache_updater

    def save_data(self, data):
        self._saver.save_data(data)
        self._mail_sender.send("admin@example.com")
        self._cache_updater.update_cache()

# Клиентский код
class DatabaseSaverClient:
    def main(self, b):
        saver = DatabaseSaverDecorator(
            DatabaseSaver(),
            MailSender(),
            CacheUpdater()
        )
        self.do_smth(saver)

    def do_smth(self, saver):
        saver.save_data(None)